pkg/registry/fallback.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright The Helm Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package registry
+
+import (
+	"crypto/tls"
+	"net/http"
+	"sync/atomic"
+)
+
+type fallbackTransport struct {
+	Base      http.RoundTripper
+	forceHTTP atomic.Bool
+}
+
+func newTransport() *fallbackTransport {
+	type cloner[T any] interface {
+		Clone() T
+	}
+	// try to copy (clone) the http.DefaultTransport so any mutations we
+	// perform on it (e.g. TLS config) are not reflected globally
+	// follow https://github.com/golang/go/issues/39299 for a more elegant
+	// solution in the future
+	baseTransport := http.DefaultTransport
+	if t, ok := baseTransport.(cloner[*http.Transport]); ok {
+		baseTransport = t.Clone()
+	} else if t, ok := baseTransport.(cloner[http.RoundTripper]); ok {
+		// this branch will not be used with go 1.20, it was added
+		// optimistically to try to clone if the http.DefaultTransport
+		// implementation changes, still the Clone method in that case
+		// might not return http.RoundTripper...
+		baseTransport = t.Clone()
+	}
+
+	return &fallbackTransport{
+		Base: baseTransport,
+	}
+}
+
+// RoundTrip wraps base round trip with conditional insecure retry.
+func (t *fallbackTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if ok := t.forceHTTP.Load(); ok {
+		req.URL.Scheme = "http"
+		return t.Base.RoundTrip(req)
+	}
+	resp, err := t.Base.RoundTrip(req)
+	if err != nil && req.URL.Scheme == "https" {
```

Comment:
Can we reference why downgrading to http is done here?

(I can speculate ORAS v1 did this, but since it is quite unusual, I think worth linking to why Helm would do this / where in ORAS v1 this behavior existed)

Just a reference in a comment here will be fine. Can PR in to source comment in a followup (try and get this merged without multiple approval cycles)

Reply:
Sure thing. 

Yes, this was only added for compatibility.

Reply:
```suggestion
	// We are falling back to http here for backward compatibility with Helm v3.
	// ORAS v1 provided fallback automatically, but ORAS v2 does not.
	if err != nil && req.URL.Scheme == "https" {
```